### PR TITLE
Open source software- and software package-friendly metadata

### PR DESCRIPTION
**Pull request by infin8x**
This PR proposes a set of changes to the JSON schema that are specifically about software, both packaged software (e.g. [npm](https://www.npmjs.com/), [pip](https://pypi.org/), etc.) and open source software (which may or may not be packaged).

 ## What value do these changes provide?

As a consumer of CVE data, GitHub would like to see these changes made because we believe it will make the CVE List's data more actionable. Specifically, this set of changes provides the necessary data for vendors like us to:

- Send alerts (e.g. GitHub's [Vulnerable Dependency Alerts](https://help.github.com/en/github/managing-security-vulnerabilities/about-security-alerts-for-vulnerable-dependencies)) about vulnerable package dependencies. These alerts generally prompt users to update to a newer version of the package that doesn't contain the vulnerability.
- Trigger automated updates (e.g. GitHub's [Automated Security Updates, formerly called Dependabot](https://help.github.com/en/github/managing-security-vulnerabilities/configuring-automated-security-updates#about-automated-security-updates)) that automatically take the action suggested by the alert in the first bullet.

We want to do both of these things at scale, so having a machine-readable, standardized format is critical for us. 

As a producer of CVE data (i.e. a CNA), GitHub already collects this data in our [Security Advisories](https://help.github.com/en/github/managing-security-vulnerabilities/about-github-security-advisories). We would like to contribute it back to the CVE List and the NVD so others can take advantage of it as well. We believe that this data is of value to anyone using or creating open source.

## Details on the changes we propose

We would like to add 2 identity fields to the JSON schema. These two identity fields let us conclusively understand the package that an advisory is about.

- `package_name`: the exact name of the package, as it appears on its package registry. For npm packages like [lodash](https://www.npmjs.com/package/lodash), we can generally get this information from the `product_name` field today. For more complexly-named package types like Maven, we cannot. Consider the example of Apache Struts. CVEs for Apache Struts generally provide a `product_name` like "Apache Struts" ([example](https://github.com/CVEProject/cvelist/blob/30847996819e0a7e6aff3bcd783d171f55d7a5bd/2017/9xxx/CVE-2017-9787.json)), which means a GitHub data curator must manually determine which packages are affected before we can trigger our alerts and updates.
- `ecosystem_name`: along with the package name, we need to know what kind of package it is. I created the current enum proposal mainly from [this list](https://en.wikipedia.org/wiki/List_of_software_package_management_systems) and my past experience building private package managers. 

We would also like to add 2 fields that can better tie a CVE to open source information.

- `repository`: the open-source repository of the product, if it is open source.
- `version_revision`: the revision indicator (e.g. a Git tag or commit hash) that is associated with a particular `version_value`. This allows us to do research on vulnerability fixes and allows independent verification of the fix by reviewing the source code at the revision that claims to fix the vulnerability.

## Remaining TODOs

- [ ] Introduce better validation rules for these newly-introduced fields

### CODE REVIEW

File: cve_json_schema/v5.x_discuss/cve507_containerized.schema [hunk 1]
```diff
@@ -49,6 +49,37 @@
...
+                        "cran",
+                        "ctan",
+                        "deb",
+                        "docker",
+                        "golang",
+                        "ivy",
+                        "maven",
+                        "npm",
+                        "nuget",
+                        "pip",
```
Comment (rschultheis) [inline cve_json_schema/v5.x_discuss/cve507_containerized.schema]: prefer `pypi` over pip personally

File: cve_json_schema/v5.x_discuss/cve507_containerized.schema [hunk 2]
```diff
@@ -49,6 +49,37 @@
...
+                    "type": "string",
+                    "description": "Name of the affected software package, if it exists."
+                },
+                "package_ecosystem": {
+                    "type": "string",
+                    "description": "If a package_name is provided, the ecosystem from which the package originates.",
+                    "enum": [
+                        "brew",
+                        "cabal",
+                        "chocolatey",
```
Comment (rschultheis) [inline cve_json_schema/v5.x_discuss/cve507_containerized.schema]: suggest adding `cargo` for rust packages

File: cve_json_schema/v5.x_discuss/cve507_containerized.schema [hunk 3]
```diff
@@ -49,6 +49,37 @@
...
+                        "composer",
+                        "conan",
+                        "conda",
+                        "cpan",
+                        "cran",
+                        "ctan",
+                        "deb",
+                        "docker",
+                        "golang",
+                        "ivy",
```
Comment (rschultheis) [inline cve_json_schema/v5.x_discuss/cve507_containerized.schema]: suggest adding `hex` for erlang/elixir packages

**Comment by infin8x** [comment]
After consulting with @chandanbn and @rschultheis, what's already in this revision for version expressions will work for us, so I've limited this PR to only the `package_name` and `package_ecosystem` additions.

File: cve_json_schema/v5.x_discuss/cve507_containerized.schema [hunk 4]
```diff
@@ -49,6 +49,39 @@
...
+                        "docker",
+                        "golang",
+                        "hex",
+                        "ivy",
+                        "maven",
+                        "npm",
+                        "nuget",
+                        "pypi",
+                        "rpm",
+                        "rubygem",
```
Comment (feelepxyz) [inline cve_json_schema/v5.x_discuss/cve507_containerized.schema]:
```suggestion
                        "rubygems",
```

**Pull request closed without merging**